Describe the hash tree construction

cole-miller opened the pull request:
This adds a formal description of how a tree is constructed by repeatedly splitting a sequence of bytes into segments. I'm sure there are some holes that need plugging, but I think I got at least the essentials right.

The first two commits in the PR are aesthetic tweaks to the LaTeX throughout the spec:

* replacing the ASCII notations `...` and `||` with LaTeX versions, `\dots` and `\mathbin{\|}` (which tend to have better spacing/layout properties)
* using the roman (non-italic) font variant for text subscripts like `max` and `min` (this improves spacing and I think is a little more readable)
* changing `\frac{x}{2^n}` to `x/2^n` (for inline fractions I find using a solidus works better than `\frac`, which is vertically cramped)

I don't mind reverting some of these if you prefer the original versions.

Closes #10.

Commit:
Second attempt to formalize the tree construction

zenhack commented:
Thanks! I merged the first two commits by themselves, since I wanted to push a fix that would have caused conflicts. I'll have a closer look at the substance of this tomorrow.

zenhack commented:
This is a decent first pass. A couple thoughts:

- It's a bit weird that we have a formal definition of something as simple as a byte, but don't have any kind of definitions prior to talking about trees and labeling nodes; perhaps we should pin down the tree terminology more explicitly at the start of the section.
- We should probably also have a definition for `TREE_C` (without the threshold subscripts) that gives you the root node of the tree, since this is the artifact that we're actually after here; you describe individual levels but not where to stop.

I also think the presentation suffers a bit due to the fact that it needs to refer to varying thresholds and the way the configuration is defined makes doing that awkward. I'm wondering if we should just scrap the configuration entirely and have separate parameters for these things. We probably should fix the window size as a constant anyway, which would bring us down to 4 parameters, which is manageable. But I think this can be addressed as a separate change.

cole-miller commented:
> perhaps we should pin down the tree terminology more explicitly at the start of the section.

Makes sense.

> We should probably also have a definition for `TREE_C` (without the threshold subscripts) that gives you the root node of the tree, since this is the artifact that we're actually after here; you describe individual levels but not where to stop.

What would `TREE_C` with no threshold sequence mean? The tree needs to know how many bits to examine at each level, and that information has to be passed into the top-level construction somehow. Were you thinking we'd decrease the threshold by one at each level, starting with the value in `C`?

> I'm wondering if we should just scrap the configuration entirely and have separate parameters for these things.

I think the configuration might make sense as an element of the highest-level presentation, but agree that for the detailed definitions using concrete parameters would be clearer.

zenhack commented:
The idea is TREE_C(X) would just be the *full* tree for the input X -- so it would have as exactly many levels as it needs to reach the root.

cole-miller commented:
Ahh, sorry, I had the wrong mental picture of the construction. Fixes on the way.

cole-miller commented:
Okay, I've rewritten the section. It took me a few tries because the most natural description of the tree construction (appearing e.g. in the Go package documentation) is bottom-up, and I wanted a recursive, top-down definition for the spec. The version I went with builds the "pruned" tree, with all the single-child nodes collapsed away, though that wouldn't be hard to change.

I ended up breaking the threshold parameter out of the configuration structure, to make the presentation easier to follow, but I left the rest of the configuration as-is for now. Finally, I've made the dependence of `I(X)` on `T` and `C` explicit, because the tree construction uses varying values of `T`.

zenhack commented on spec.md:
```diff
@@ -64,42 +64,72 @@ We also use the following operators:
+is a leaf node will also be referred to as a "leaf".
+
+Let $C$ be a configuration, and let $T$ and $S$ be positive integers. We
+call $T$ the "threshold", as before, and $S$ the "step". Now for $X \in
```
Nitpick, but I think the choice of `S` as a variable name here is troublesome in that it might suggest a relationship with S_min and S_max, though no such relationship exists.

cole-miller replied:
Good point, I'll pick something else.

zenhack commented on spec.md:
```diff
@@ -64,42 +64,72 @@ We also use the following operators:
+  whose children in order are the subtrees $\operatorname{TREE}_{T, S,
+  C} (Y)$ for each entry $Y$ in $\operatorname{SPLIT}_{T + nS, C} (X)$,
+  where $n$ is the largest integer such that $I_{T + nS, C} (X)$
+  exists.
```
If I'm reading this correctly, I don't think this construction quite works as intended: S_max is going to bound the maximum size of a chunk, so you can't just apply split with a greater T but the same C and expect it to group the smaller chunks. 

I think this is the right way to approach this conceptually, but that interaction needs fixing.

cole-miller replied:
You're quite right, though it took me a little while to see it. I have a fix in mind, will update once it's written up.

zenhack reviewed (changes requested):
This looks much better. A couple of issues in-line, and also you've got merge conflicts with stuff on master that need fixing.

Commit:
Replace parameter name: S -> B

This avoids a clash with the unrelated parameters S_min and S_max.

Commit:
Fix ill-definedness of TREE with maximal-size chunks

Also a couple of other problems:

- The base case for the TREE recursion should fire when I_{T, C} (X)
  *does not* exist, not when it *does*
- TREE_{T, B, C} (X) doesn't have an "inorder" traversal, because it's
  not a BST! Specify the depth-first traversal instead.

cole-miller commented:
Third try: this should fix the unboundedness problem pointed out by @zenhack, and also corrects some other errors I noticed in the process of rewriting (see the commit message 4faf2ed). The involvement of `S_max` just makes everything a bit more fiddly. Perhaps I should implement both this recursive algorithm and the bottom-up version in Haskell or something and check that they give the same results.

(Waiting to fix the conflicts until this is ready to merge.)

zenhack commented:
You should probably go ahead and fix the conflicts; they're not entirely mechanical as a few definitions changed on master. In particular, I(X) is always defined (the way it was defined before had a mistake re: max block size), so that will require some reframing. I'll have another look at this after dinner.

Commit:
Update TREE to reflect new definition of I(X)

zenhack commented on spec.md:
```diff
@@ -72,41 +72,70 @@ We also use the following operators and functions:
+  positive integer exists. Then $\operatorname{TREE}_{T, B, C} (X)$ is a
+  node labelled by $X$ whose children in order are the subtrees
+  $\operatorname{TREE}_{T, B, C} (Y)$ for each entry $Y$ of
+  $\operatorname{SPLIT}_{T + nB, C} (X)$.
```
I think this still has the problem with S_max: In the recursive case, simply shelling out to `SPLIT_{T + nB, C}` doesn't necessarliy give you a grouping of the chunks produced by `SPLIT_{T, C}`.

For example: if `SPLIT_{T, C}` makes it's first split at position `S_max - 4`, and its second at `S_max + 12`, assuming the first split didn't have any more bits set than necessary, `SPLIT_{T + nB, C}` will choose its first index as `S_max`, since it will hit that before the next viable split point. But that was *not* a split point with the smaller threshold. So you end up with different leaves, not just more of them.

cole-miller replied:
(Apologies for the delay in responding!) I guess I missed the point of your previous review, so thanks for spelling it out here. I agree that the algorithm has the property you describe, but I don't see why it's problematic -- does it interfere with the incremental recomputation procedure, or cause the constructed tree to differ from the bottom-up version, or something else? I am pretty sure that this step is at least well-defined, i.e. there exists a largest `n` satisfying the condition, because we can always pick an integer so large that the rolling hash never has enough trailing zeroes within our chunk and hence the only splits are caused by hitting the maximum size.

zenhack replied:
It means it fails the last property in your list below:

> The labels of the leaf nodes in their natural order (inherited from the depth-first traversal of the tree) are exactly the entries of $\operatorname{SPLIT}_{T, C} (X)$.

...which I think is important: the idea is to arrange the chunks into a tree; it should not modify the split points of the chunks themselves.

(It does also differ from the bottom-up implementation, in that the bottom-implementation maintains that property).

cole-miller replied:
Oh, duh. Okay, it seems like we can fix that by waiting to apply the `S_max` cutoff: we go all the way down to `T + 0B` using just the hash-based splitting, and only then do we break down any remaining chunks bigger than `S_max`. That reflects how the bottom-up algorithm proceeds, only applying the cutoff if no "conventional" split intervenes first.

So the recursive definition will require a version of `SPLIT` that ignores `S_max` -- I'll edit that section accordingly. And since this approach has proved to be a bit tricky, it would be nice to include the bottom-up, iterative definition as well, though that probably belongs in a separate PR.

zenhack replied:
I'm not sure that quite works either. Suppose the first natural split point is at position `S_max + N` where `N < S_min`. Because we hit `S_max` first, we'll split there, and then we will reject `S_max + N` as a split point because it results in a chunk that is too small. So you can't just split everything on natural boundaries and then add extra splits for `S_min`/`S_max` after the fact, since doing so can also require you to reject natural split points, not just add new split points.

I wonder if it is better to just use the bottom-up definition. Some arguments in favor:

- It lines up more closely to what implementations will do anyway.
  - We will want to document the bottom-up algorithm either way, since the correspondence is not entirely obvious.
- It's symmetric with what `SPLIT` does.

Especially if we start exploring things like min/max branching factors later, the symmetry with `SPLIT` is likely to be an asset.

I think before getting into the trenches the top-down definition seems more elegant & declarative, but I'm not sure what it buys us pragmatically, given how fiddly it has been to state it correctly; maybe just having the list of properties that hold is good enough.

cole-miller replied:
I'd thought `S_min` couldn't cause problems but you're quite right, it does. And the complexity of coalescing nodes at the end to repair the lower bound, and how that interacts with the fix for `S_max`, does seems like reason enough to abandon the top-down approach (certainly you'd never implement it this way, even in a purely functional language). I'll try writing up the bottom-up definition tonight.

bobg commented:
I started working on my own language for tree construction, before I was aware that you were already doing this work. For what it's worth, and in case it helps, here's the gist of what I started writing (based on the implementation in github.com/bobg/hashsplit):

- The output of the split algorithm is a sequence of "chunks,"
- If H was the value of the rolling checksum when a chunk K was made, then the "level" L(K) is Q-T, where Q is the largest integer <= 32 such that H mod 2^Q = 0 (i.e., the level is the number of trailing zeroes in the rolling checksum in excess of the threshold)
- A node in a hashsplit tree also has a level
- A level-0 node has chunks for children
- A node at level N > 0 has children that are nodes at level N-1

To build a tree, start with an empty level-0 node. This is "the current level 0 node." Then for each chunk emerging from the split algorithm:

- Add the chunk to the current level-0 node.
- For each integer i in the range [0 .. L), "close" the current level-i node

To "close" a node at level L:

- If there is no node at level L+1, create one. This is "the current level L+1 node."
- Add the closing level-L node as a child to the current level L+1 node.
- Create a new level-L node. This is now "the current level L node."

After all chunks are added to the tree in this way:

- If the current level-0 node has any children, "close" it, and "close" each "current" node above it in turn.
- The highest-level "current" node is the _prospective_ root of the tree.
- [pruning] While the prospective root has only one child, set the prospective root to that child.
- The prospective root is the root of the final tree.

zenhack commented:
@cole-miller, are you still interested in working on this?

cole-miller commented:
Sorry for dropping the ball here; I'm happy to let @bobg take over.
